docs/plugins/testing_workshop_docs/Article-3-readers-and-fixtures.md

```diff
@@ -0,0 +1,152 @@
+  
+Whatever is returned out of a `@pytest.fixture` decorated function is passed as an argument with the name of the fixture, to the test. We are going to call this `pytest.fixture` decorated function `write_im_to_file`, as defined on line 6. We’re going to give this fixture the `tmp_path` fixture. Fixtures can use fixtures. 
+
+`write_im_to_file` returns a function we will call `write_func` that we can pass a path to and have it write the numpy file to that path. `write_func` is defined inside `write_im_to_file` because it’s not needed anywhere else. When we use `write_im_to_file` inside this test it will be referred to as `write_func`. We can call it with a file name, save the data, and read the data back.   
+  
+We will have access to what `write_func` returns once it’s been called inside the test. It returns both the path where the data has been written and the original data. It accepts an actual file name that we are going to pass.   
+  
+The benefit of creating this fixture is that whenever we want to write our own test data we don't have to copy three lines of code, we can write just one long line of code. This is useful in testing data with different structures like integers or a specific layer type. Those arguments could be passed to further customize your fixture.  
+  
+We still want to make sure we get a reader when we call `napari_get_reader` with the file. We call that `reader` function with the file we created to see if it returns what we expect. Remember the [reader spec](https://napari.org/stable/plugins/contributions.html#contributions-readers), it should actually return a layer data list. Here is the full test, with the fixture:  
+```python   
+    @pytest.fixture  
+    def write_im_to_file(tmp_path):  
+ 
+        def write_func(filename):  
+            my_test_file = str(tmp_path / filename)  
+            original_data = np.random.rand(20, 20)  
+            np.save(my_test_file, original_data)  
+       
+            return my_test_file, original_data  
+ 
+        return write_func  
+   
+    def test_reader_round_trip(write_im_to_file):  
+        my_test_file, original_data = write_im_to_file("myfile.npy")  
+     
+        reader = napari_get_reader(my_test_file)  
+        assert callable(reader)  
+     
+        layer_data_list = reader(my_test_file)  
+        assert isinstance(layer_data_list, List) and len(layer_data_list) > 0  
+     
+        layer_data_tuple = layer_data_list[0]  
+        layer_data = layer_data_tuple[0]  
+        np.testing.assert_allclose(layer_data, original_data)  
+``` 
+We’re going to assert a list length greater than zero. There must be a layer in there; otherwise, we didn't read it correctly. We also assert that it is a list. Once we start testing, there are many things we could test. Try to focus on what's most useful.  
```

Comment:
```suggestion
We’re going to assert a list length greater than zero. There must be a layer in there; otherwise, we didn't read it correctly. We also assert that it is a list.
```